java/client/src/main/java/io/vitess/client/VTGateConnection.java

```diff
@@ -0,0 +1,203 @@
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.vitess.client.cursor.Cursor;
+import io.vitess.client.cursor.CursorWithError;
+import io.vitess.client.cursor.SimpleCursor;
+import io.vitess.client.cursor.StreamCursor;
+import io.vitess.proto.Query;
+import io.vitess.proto.Query.SplitQueryRequest.Algorithm;
+import io.vitess.proto.Vtgate;
+import io.vitess.proto.Vtgate.*;
+
+import javax.annotation.Nullable;
+import java.io.Closeable;
+import java.io.IOException;
+import java.sql.SQLDataException;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.util.concurrent.Futures.transformAsync;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+
+public final class VTGateConnection implements Closeable {
+    private final RpcClient client;
+    private Session session;
+    private SQLFuture<?> lastCall;
+
+    public VTGateConnection(RpcClient client) {
+        this(client, null, null);
+    }
+
+    public VTGateConnection(RpcClient client, String targetString) {
+        this(client, targetString, null);
+    }
+
+    public VTGateConnection(RpcClient client, Query.ExecuteOptions options) {
+        this(client, null, options);
+    }
+
+    public VTGateConnection(RpcClient client, String targetString, Query.ExecuteOptions options) {
+        this.client = checkNotNull(client);
+        this.session = Session.newBuilder()
+                .setTargetString(targetString)
+                .setOptions(options)
+                .build();
+    }
+
+    public SQLFuture<Cursor> execute(Context ctx, String query, @Nullable Map<String, ?> bindVars) throws SQLException {
+        checkCallIsAllowed("execute");
+        ExecuteRequest.Builder requestBuilder = ExecuteRequest.newBuilder()
+                .setQuery(Proto.bindQuery(checkNotNull(query), bindVars))
+                .setSession(session);
+
+        if (ctx.getCallerId() != null) {
+            requestBuilder.setCallerId(ctx.getCallerId());
+        }
+
+        SQLFuture<Cursor> call = new SQLFuture<>(
+                transformAsync(
+                        client.execute(ctx, requestBuilder.build()),
+                        new AsyncFunction<ExecuteResponse, Cursor>() {
+                            @Override
+                            public ListenableFuture<Cursor> apply(ExecuteResponse response) throws Exception {
+                                setSession(response.getSession());
+                                Proto.checkError(response.getError());
+                                return Futures.<Cursor>immediateFuture(new SimpleCursor(response.getResult()));
+                            }
+                        },
+                        directExecutor()));
+        lastCall = call;
```

Comment:
I see you've synchronized `checkCallIsAllowed` below, but I dont think that's enough since lastCall can be set from any of these functions